src/tools/rbd_mirror/ImageDeleter.cc

```diff
@@ -0,0 +1,367 @@
+
+      if (!m_running.read()) {
+        m_delete_queue_lock.Unlock();
+        dout(20) << "return" << dendl;
+        return;
+      }
+    }
+
+    unique_ptr<DeleteInfo> del_info = std::move(m_delete_queue.front());
+    m_delete_queue.pop();
+    m_delete_queue_lock.Unlock();
+
+    process_image_delete(std::move(del_info));
+  }
+}
+
+void ImageDeleter::schedule_image_delete(uint64_t local_pool_id,
+                                         const std::string& local_image_id,
+                                         const std::string& global_image_id) {
+  dout(20) << "enter" << dendl;
+  Mutex::Locker l(m_delete_queue_lock);
+  m_delete_queue.push(unique_ptr<DeleteInfo>(
+        new DeleteInfo(local_pool_id, local_image_id, global_image_id)));
+  m_delete_queue_cond.Signal();
+}
+
+void ImageDeleter::process_image_delete(unique_ptr<DeleteInfo> del_info,
+                                        FailedStep step, int retries) {
+
+  std::string del_info_str = del_info->to_string();
+  dout(20) << "start processing delete request: " << del_info_str << dendl;
+  int r;
+  cls::rbd::MirrorImage mirror_image;
+
+  // remote image was disabled, now we need to delete local image
+  IoCtx ioctx;
+  r = m_local->ioctx_create2(del_info->local_pool_id, ioctx);
+  if (r < 0) {
+    derr << "error accessing local pool: " << cpp_strerror(r) << dendl;
+    enqueue_failed_delete(std::move(del_info), FAILED_ON_DISABLE_MIRRORING, r,
+                          retries);
+    return;
+  }
+
+  bool has_snapshots;
+  bool has_children;
+  r = image_has_snapshots_and_children(&ioctx, del_info->local_image_id,
+                                       &has_snapshots, &has_children);
+  if (r < 0) {
+    enqueue_failed_delete(std::move(del_info), FAILED_ON_DISABLE_MIRRORING, r,
+                          retries);
+    return;
+  }
+
+  if (has_children) {
+    // there are still clones of snapshots of this image, therefore send the
+    // delete request to the end of the queue
+    dout(20) << "local image id " << del_info->local_image_id << " has "
+             << "snapshots with cloned children, postponing deletion..."
+             << dendl;
+    Mutex::Locker l(m_delete_queue_lock);
+    m_delete_queue.push(std::move(del_info));
+    return;
+  }
+
+  string img_name;
+  r = cls_client::dir_get_name(&ioctx, RBD_DIRECTORY, del_info->local_image_id,
+                               &img_name);
+  if (r < 0) {
+    derr << "error retrieving local image name for image id "
+         << del_info->local_image_id << ": " << cpp_strerror(r) << dendl;
+    enqueue_failed_delete(std::move(del_info), FAILED_ON_DISABLE_MIRRORING, r,
+                          retries);
+    return;
+  }
+
+  switch (step) {
+    case FAILED_ON_REMOVE_SNAPS:
+      goto failed_on_remove_snaps;
+    case FAILED_ON_REMOVE_IMAGE:
+      goto failed_on_remove_image;
+    case FAILED_ON_DISABLE_MIRRORING:
+      break;
+  }
+
+  mirror_image.global_image_id = del_info->global_image_id;
```

Comment:
Yeah -- I think it can only be a bug for -EEXIST.  In such a case, log it and abort seems reasonable for now.  I think -EINVAL is possible if the user force-promotes the local image, disables mirroring, and re-enables it (e.g. crazy user).  Again, I think log and abort seems reasonable.